packages/amplify-category-api/src/provider-utils/awscloudformation/PipelineWithAwaiter.ts

```diff
@@ -0,0 +1,380 @@
+            lambda: (() => {
+              const preDeployCodeFilePath = path.join(__dirname, 'lambdas', 'predeploy.js');
+              const lambdaHandlerCode = fs.readFileSync(preDeployCodeFilePath, 'utf8');
+
+              const action = new lambda.Function(scope, 'PreDeployLambda', {
+                code: lambda.InlineCode.fromInline(lambdaHandlerCode),
+                handler: 'index.handler',
+                runtime: lambda.Runtime.NODEJS_12_X,
+                environment: {
+                  DESIRED_COUNT: `${desiredCount}`,
+                  CLUSTER_NAME: service.cluster,
+                  SERVICE_NAME: service.serviceName,
+                },
+              });
+
+              action.addToRolePolicy(
+                new iam.PolicyStatement({
+                  actions: ['ecs:UpdateService'],
+                  effect: iam.Effect.ALLOW,
+                  resources: [cdk.Fn.ref(service.logicalId)],
+                }),
+              );
+
+              return action;
+            })(),
+            inputs: [],
+            outputs: [],
+          }),
+        ],
+      },
+      {
+        stageName: 'Deploy',
+        actions: [
+          new codepipelineactions.EcsDeployAction({
+            role: getRole(scope, 'UpdateDeploy'),
+            actionName: 'Deploy',
+            service: new (class extends cdk.Construct implements ecs.IBaseService {
+              cluster = {
+                clusterName: service.cluster,
+                env: {},
+              } as ecs.ICluster;
+              serviceArn = cdk.Fn.ref(service.serviceArn);
+              serviceName = service.serviceName;
+              stack = cdk.Stack.of(this);
+              env = {} as any;
+              node = service.node;
+            })(this, 'tmpService'), // TODO: clean this,
+            input: buildOutput,
+          }),
+        ],
+      },
+    ]);
+
+    const role = getRole(scope, `Pipeline`, new iam.ServicePrincipal('codepipeline.amazonaws.com'));
+
+    const pipeline = new codepipeline.Pipeline(scope, `${id}Pipeline`, {
+      pipelineName: `codepipeline-amplify-${service.serviceName}`,
+      crossAccountKeys: false,
+      artifactBucket: bucket,
+      stages: stagesWithDeploy,
+      role,
+    });
+
+    pipeline.node.addDependency(service);
+
+    const pipelineAwaiter = new PipelineAwaiter(scope, 'Awaiter', {
+      pipeline,
+      artifactBucketName: bucket.bucketName,
+      artifactKey: s3SourceActionKey,
+      deploymentMechanism,
+    });
+  }
+}
+
+function createPreBuildStages(
+  scope: cdk.Construct,
+  {
+    bucket,
+    s3SourceActionKey,
+    githubSourceActionInfo,
+    sourceOutput,
+    roleName,
+  }: {
+    bucket: s3.IBucket;
+    s3SourceActionKey: string;
+    githubSourceActionInfo?: GitHubSourceActionInfo;
+    sourceOutput: codepipeline.Artifact;
+    roleName: string;
+  },
+) {
+  const stages: codepipeline.StageOptions[] = [];
+
+  const stage = {
+    stageName: 'Source',
```

Comment:
is stageName equivalent to environment name? Should it be always hardcoded to "Source"?

Reply:
What do you mean by "environment name"?